### PR TITLE
Run tests on fresh Linux environments

### PR DESCRIPTION
**Pull request by rs--**
Install system dependencies on Linux (RHEL 8, Ubuntu 16.04/18.04) before building, and testing OE. Windows will be added at a later date.

### CODE REVIEW

**Comment by rs--** [comment]
bors try

**Comment by rs--** [comment]
bors try

**Comment by jhand2** [comment]
For posterity, can you add to the PR description/commit message what you mean by end-to-end tests (that is, explain that it also tests installation of packages)

**Comment by anakrish** [comment]
@rs-- @BRMcLaren Are we doing e2e testing on every job/PR?

**Comment by rs--** [comment]
@anakrish Yeah, the plan is to run it on a few configurations for every PR

**Comment by anakrish** [comment]
Do you think it would be better to run e2e tests as periodically say twice a day instead of on every PR?

The reason I ask is that, the scripts change infrequently compared to code and the chance of a typical PR breaking the scripts is minimal, though not zero. The URLs that the script depends on however change occasionally, breaking the powershell script. So it seems periodic jobs would help identify a remote url change.

Some of the issues the scripts have occur due to reuse in an existing system:
E.g: #3043 #3044

Is there another strategy for catching those errors?

**Comment by BRMcLaren** [comment]
> Do you think it would be better to run e2e tests as periodically say twice a day instead of on every PR?

FWIW there are e2e tests running in the background that are not reported up to the repo as badges. It's not 100% stable yet, but for the [curious](https://oe-jenkins-tf.westeurope.cloudapp.azure.com/job/CI-CD_Infrastructure/job/OpenEnclave-Rebuild-Infrastructure/)

Regarding your suggestion, I don't believe that is the [issue](https://github.com/openenclave/openenclave/issues/2993) @rs--  is trying to solve. Part of the current process of integration when an individual modifies scripts is sending them through a rebuild of the above Rebuild Infrastructure job... twice, which takes 16 or so hours and someone with Jenkins permissions to assist. It would be better for developers if there was an automated way for them to contribute to the project, add packages if they need to, fix broken URLs, or bugs to encourage independence.

I would add that having a subset of the testing configuration run on new machines to test the full Contributor experience seems nice and would guard against regressions. It would also allow developers to try new things and have them caught during integration time, things such as if someone wanted to use a new package manager instead of the ps1 script, the framework this PR implements would support making that extendable.

**Comment by BRMcLaren** [comment]
> Is there another strategy for catching those errors?

This is an interesting proposition, maybe an e2e test that pulls from old machine images/definitions and runs as part of the maintenance testing instead of per PR?

**Comment by anakrish** [comment]
> > Do you think it would be better to run e2e tests as periodically say twice a day instead of on every PR?
> 
> FWIW there are e2e tests running in the background that are not reported up to the repo as badges. It's not 100% stable yet, but for the [curious](https://oe-jenkins-tf.westeurope.cloudapp.azure.com/job/CI-CD_Infrastructure/job/OpenEnclave-Rebuild-Infrastructure/)
> 
That is great! Hopefully that will catch OpenSSL/Intel urls changing ahead of time.

**Comment by anakrish** [comment]
>  It would be better for developers if there was an automated way for them to contribute to the project, add packages if they need to, fix broken URLs, or bugs to encourage independence.

This is good. I am planning to remove node/esy and add native ocaml install to both the windows and linux scripts. So it will be nice any mistakes are caught beforehand.

**Comment by anakrish** [comment]
> > Is there another strategy for catching those errors?
> 
> This is an interesting proposition, maybe an e2e test that pulls from old machine images/definitions and runs as part of the maintenance testing instead of per PR?

That would work. That kind of test does not have to run every PR. The script for example does lots of things like uninstall old-driver, install new one, install visual studio etc.
I personally don't see value in testing all these scenarios on each PR - very few PRs have changes that will affect the script. However these are scenarios that ought to be locked down nonetheless.

File: .jenkins/pipelines/Azure/Linux/Jenkinsfile [hunk 1]
```diff
@@ -16,14 +16,21 @@ AGENTS_LABELS = [
     "ubuntu-nonsgx":    env.UBUNTU_NONSGX_CUSTOM_LABEL ?: "nonSGX",
     "windows-nonsgx":   env.WINDOWS_NONSGX_CUSTOM_LABEL ?: "nonSGX-Windows"
 ]
+VANILLA_AGENTS_LABELS = [
```
Comment (BRMcLaren) [inline .jenkins/pipelines/Azure/Linux/Jenkinsfile:19]: I am wondering if it is best to have two different structures VANILLA and the other, as opposed to a single lookup map.

Comment (rs--) [reply .jenkins/pipelines/Azure/Linux/Jenkinsfile:19]: An option would be to have it in the same map, but just use `acc-rhel-8-vanilla`?

**Comment by rs--** [comment]
bors try

File: .jenkins/pipelines/Azure/Linux/Jenkinsfile (hunk 1 above)
Comment (BRMcLaren) [reply .jenkins/pipelines/Azure/Linux/Jenkinsfile:19]: Yes, this is what I was thinking.

**Comment by BRMcLaren** [comment]
@rs--  Whenever we add complicated improvements like this, it's good to see stability. I have approved the change, but it would be good to see having 5 or so successful runs in a row to ensure this is not introducing too many failure points.

**Comment by rs--** [comment]
bors try

**Comment by rs--** [comment]
bors try

**Comment by rs--** [comment]
bors try

**Comment by rs--** [comment]
bors try

**Comment by BRMcLaren** [comment]
bors try

**Comment by BRMcLaren** [comment]
bors try

**Comment by rs--** [comment]
bors try

**Comment by BRMcLaren** [comment]
bors try

**Pull request merged**